modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/ColocationGroup.java

```diff
@@ -228,6 +228,27 @@ public ColocationGroup finalaze() {
         return forNodes0(nodeIds);
     }
 
+    /** */
+    public ColocationGroup filterForPartitions(int[] parts) {
+        if (assignments != null) {
+            List<List<UUID>> assignments = new ArrayList<>(this.assignments.size());
+            Set<UUID> nodes = new HashSet<>();
+
+            for (int i = 0; i < this.assignments.size(); ++i) {
+                UUID first = F.isEmpty(parts) || Arrays.binarySearch(parts, i) >= 0 ? F.first(this.assignments.get(i)) : null;
+
+                if (first != null)
+                    nodes.add(first);
+
+                assignments.add(first != null ? Collections.singletonList(first) : Collections.emptyList());
```

Comment:
Assignment should already be finilized and can contain only 1 value in the list. We can use original `assignments.get(i)` instead of building new singleton list (reduce GC pressure a little bit).